### PR TITLE
devicetree: allow interating devicetree nodes with identical zephyr,device-type properties

### PR DESCRIPTION
**Pull request by henrikbrixandersen**
This PR add a base zephyr,device-type property for indicating the Zephyr-specific type of device represented by a given binding. Knowing the type of device will allow iterating through a devicetree nodes with the same type at build-time, an often requested feature.

As examples, the base ADC controller and base GPIO controller devicetree bindings have been updated to make use of this new property. If this is accepted, all devicetree base bindings should have this added.

Apart from the tests cases, the list of hardcoded compatibles in the ADC shell module has been replaced with a one-liner, iterating all ADC controller devicetree nodes. This not only saves us having to update this list, but it also makes the ADC shell work with out-of-tree ADC controller bindings/drivers.

This functionality can be used many places in the tree.

### CODE REVIEW

**Review by gmarull — COMMENTED** [review]
I have to say that I'm not convinced by this approach. This is yet another DT property that has nothing to do with hardware description. Zephyr's flavor of DT should soon deserve its own name (or standard?), as it has deviated so much from Linux DT. If we need to iterate over devices based on their class, what I think we need are device class specific instantiation macros, so that we can store devices in multiple lists (using iterable sections). This would likely allow to also have N class instances of a single DT node, provided we have getter macros that are class specific, ie, DEVICE_GPIO_DT_GET() -> __device_gpio_dts_ord_N.

**Comment by henrikbrixandersen** [comment]
@gmarull Yeah, I know that this does have a slight hacky feel to it. Thanks for looking through it. 

It was an idea I had while working in a different issue, and I couldn't really let it go without giving it a try. 

I'd be interested in hearing other opinions too.

**Comment by bjarki-trackunit** [comment]
The last time I proposed something similar, namely adding an `api` property to the devicetree, it was argued that the API is not a hardware property. I know this PR only mentions device class, but I'm guessing we expect the API to match the device type as well? In any case, I pointed to the deprecated `device_type` property (according to DT spec v0.4) as an argument for having the type in the devicetree, which was not exactly met with open arms :D (very much no to that).

I like the approach of the macros like `SENSOR_DEVICE_DT_DEFINE` which is quite simple, and I may want to do the same for RTCs (to be able to automatically get time using standard time functions like `mktime`, although `zephyr,rtc = &rtc` may be better for that).

RTCs are actually a good example of where the API does not match the device type, given that some RTCs in-tree can implement either the counter API or the RTC API depending on if CONFIG_COUNTER=y or CONFIG_RTC=y, in this case, the sensor type macro will more accurately tag the device as an RTC or counter than the `zephyr,devicetype` property. I know its an outlier, but it's notable.

That said, how about the required properties used for `gpio-controller` and `interrupt-controller`? These seem to be able to be used in place of the `zephyr,devicetype`? should we add this to ADCs, like `adc-controller`?

**Comment by nordicjm** [comment]
> If we need to iterate over devices based on their class, what I think we need are device class specific instantiation macros, so that we can store devices in multiple lists (using iterable sections).

The device specific class macro idea sounds like a lot of work for someone creating their own drivers classes out of tree, they need to edit/add linker files (so basically have to fork zephyr), add device instance macros that presumably would need to reference some unique ID - and how would that ID remain unique? If they then updated to the next zephyr release which adds e.g. 2 extra driver classes in tree, they'd then have a collision? Or would there be no IDs at all and just names, but then how would you iterate over all devices (ignoring types)?

**Comment by henrikbrixandersen** [comment]
> That said, how about the required properties used for `gpio-controller` and `interrupt-controller`? These seem to be able to be used in place of the `zephyr,devicetype`? should we add this to ADCs, like `adc-controller`?

This was actually my initial approach, as I already use the `gpio-controller` property to iterate all GPIO controller nodes in the GPIO hogs driver:

https://github.com/zephyrproject-rtos/zephyr/blob/574e641297b1d97bb43d0cb0ba985a058fe53e66/drivers/gpio/gpio_hogs.c#L38

**Comment by henrikbrixandersen** [comment]
> The device specific class macro idea sounds like a lot of work for someone creating their own drivers classes out of tree, they need to edit/add linker files (so basically have to fork zephyr)

I share that concern. While I agree that the introduced `zephyr,device-type` property doesn't really describe hardware (although it kind of does, similar to the `gpio-controller` property), the approach here supports out-of-tree driver classes without anything else needed than three lines in an out-of-tree YAML file.

By using `default` and `const` in base bindings, the property will only be part of the generated `zephyr.dts` and should never have to appear in user-written DTS files, which somewhat encapsulates this.

**Comment by Laczen** [comment]
This proposal looks like a re-introduction of `device_type` that has been deprecated.

The `compatible` construction has the same functionality: create an umbrella compatible and use this to walk. The device_type would just replicate this construct with a lot of source for discussion: what device_types will be allowed, should multiple device types be supported, ...

**Comment by henrikbrixandersen** [comment]
Rebased on `main`.

**Comment by bjarki-trackunit** [comment]
Will it be possible to define two types for a single node? gpio-controllers can also be interrupt-controllers, which is possible with the dts properties `gpio-controller` and `interrupt-controller` See [intc-gpio test binding](https://github.com/zephyrproject-rtos/zephyr/blob/466cd14fb6b6ac7aa83990faee58ebb9f7a7136c/dts/bindings/test/vnd%2Cgpio-intc-device.yaml) and [BCM11130 linux binding](https://www.kernel.org/doc/Documentation/devicetree/bindings/gpio/brcm%2Ckona-gpio.txt) for example :)

I'm guessing zephyr,type will refer to the zephyr API used for the device, so a gpio/intc node would get the zephyr device type gpio-controller?

**Comment by henrikbrixandersen** [comment]
> Will it be possible to define two types for a single node? gpio-controllers can also be interrupt-controllers, which is possible with the dts properties `gpio-controller` and `interrupt-controller` See [intc-gpio test binding](https://github.com/zephyrproject-rtos/zephyr/blob/466cd14fb6b6ac7aa83990faee58ebb9f7a7136c/dts/bindings/test/vnd%2Cgpio-intc-device.yaml) and [BCM11130 linux binding](https://www.kernel.org/doc/Documentation/devicetree/bindings/gpio/brcm%2Ckona-gpio.txt) for example :)

We could extend this to support a string-array of device types, but since - in Zephyr - each devicetree node, at least currently, maps to only one device type, I don't see a need for it yet.

If this is needed in the future, the APIs can stay the same, only the implementation and the property type needs to be updated. 

> I'm guessing zephyr,type will refer to the zephyr API used for the device, so a gpio/intc node would get the zephyr device type gpio-controller?

Exactly. A GPIO controller supporting pin interrupts is still a GPIO controller in this regard.

**Comment by bjarki-trackunit** [comment]
What do you think of this solution?
https://github.com/zephyrproject-rtos/zephyr/blob/45205f865e18decbb9568c4691686dc679cbe961/include/zephyr/drivers/sensor.h#L1317-L1324

It seems to solve a similar issue, at least for the shell case.

**Comment by henrikbrixandersen** [comment]
> What do you think of this solution?
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/45205f865e18decbb9568c4691686dc679cbe961/include/zephyr/drivers/sensor.h#L1317-L1324

The "sensor info" solution has a run-time cost of added flash usage. The solution in this PR has no runtime cost.

**Comment by fabiobaltieri** [comment]
@henrikbrixandersen @bjarki-trackunit have you considered generating some array with a list of devices by API type based on the API struct name instead? Ultimately what we really want is all devices that implement a certain API, maybe we could have a script that goes through all the objects looking for device structures, group them by the api struct name that they implement and generate some source from it. Then one would not have to set the type manually as a free string.

**Comment by fabiobaltieri** [comment]
@pdgendt I was thinking just normal arrays for this one, that way they'd get discarded automatically if not used, I'd imagine most of the time they won't

It sounds a bit complex though, first one would have to generate the extern for the array pointer and size based on all defined api types, then build the object, then generate the file with the array list and size and build that as well, then link.

**Comment by fabiobaltieri** [comment]
@pdgendt actually yeah I think your suggestion makes sense, the API struct are in rodata right now right? There's nothing special to them? So one could think about putting them in sections by type and then iterate over them, it'd be a bit simpler than what I was thinking. The drawback is that then you actually want the struct device so you'd have to scan the device list to find which one correspond to the API pointer but that should not be a big deal (what I had in mind was to resolve that at build time and to a list of struct device pointers).

**Comment by pdgendt** [comment]
> @pdgendt actually yeah I think your suggestion makes sense, the API struct are in rodata right now right? There's nothing special to them? So one could think about putting them in sections by type and then iterate over them, it'd be a bit simpler than what I was thinking. The drawback is that then you actually want the struct device so you'd have to scan the device list to find which one correspond to the API pointer but that should not be a big deal (what I had in mind was to resolve that at build time and to a list of struct device pointers).

Yeah, there would be some runtime overhead to iterate the devices instead of the APIs, but there won't be additional ROM overhead.
I think it could coexist with what is proposed in this PR, but maybe it's undesired to have both.

**Comment by pdgendt** [comment]
An additional benefit of using the linker sections, is that you can add type checking in the API calls.

```c
static inline int z_impl_spi_transceive(const struct device *dev,
					const struct spi_config *config,
					const struct spi_buf_set *tx_bufs,
					const struct spi_buf_set *rx_bufs)
{
	const struct spi_driver_api *api =
		(const struct spi_driver_api *)dev->api;
	int ret;
	STRUCT_SECTION_START_EXTERN(spi_driver_api);
	STRUCT_SECTION_END_EXTERN(spi_driver_api);

	__ASSERT_NO_MSG(STRUCT_SECTION_START(spi_driver_api) <= api);
	__ASSERT_NO_MSG(api < STRUCT_SECTION_END(spi_driver_api));
	/* Optionally add a runtime check */

	ret = api->transceive(dev, config, tx_bufs, rx_bufs);
	spi_transceive_stats(dev, ret, tx_bufs, rx_bufs);

	return ret;
}
```

**Comment by fabiobaltieri** [comment]
> An additional benefit of using the linker sections, is that you can add type checking in the API calls

I was just thinking the same :-)  also I think you could use

```c
__ASSERT_NO_MSG(IN_RANGE(api, STRUCT_SECTION_START(spi_driver_api), STRUCT_SECTION_END(spi_driver_api));
```

in that snippet, may even make sense to have a macro for that in `iterable_sections.h`.

**Comment by Laczen** [comment]
@pdgendt, @fabiobaltieri, the creation of linker-sections or run-time access is already supported by the build system.

When a `compatible` describing the device-type is added to the dts-node it is possible to just add the linker sections based upon `#define DT_DRV_COMPAT xxx` and `DT_INST_FOREACH_STATUS_OKAY` or in the driver code validate the existence of the "device-type compatible" and add it to allow run-time validation.

IMO this is why the device-type has been deprecated in devicetree as it is easily replaced by a compatible.

**Comment by bjarki-trackunit** [comment]
I made an attempt at making sorting by devicetype at build time using macros like dts a while back #58112 It's a bit more complicated as we are generating headers at different build stages, but the underlying idea was the same :)

**Comment by pdgendt** [comment]
> When a `compatible` describing the device-type is added to the dts-node it is possible to just add the linker sections based upon `#define DT_DRV_COMPAT xxx` and `DT_INST_FOREACH_STATUS_OKAY` or in the driver code validate the existence of the "device-type compatible" and add it to allow run-time validation.

This would create an additional array/structure, whereas placing the existing API structs into a linker section has no additional ROM overhead.

**Comment by Laczen** [comment]
> > When a `compatible` describing the device-type is added to the dts-node it is possible to just add the linker sections based upon `#define DT_DRV_COMPAT xxx` and `DT_INST_FOREACH_STATUS_OKAY` or in the driver code validate the existence of the "device-type compatible" and add it to allow run-time validation.
> 
> This would create an additional array/structure, whereas placing the existing API structs into a linker section has no additional ROM overhead.

Adding them directly into a linker section could also be done, the compatible is describing the name of the linker section, a "registration" to this linker section is possible with a construct like the one utilized in registering static settings in the settings subsystem.

**Comment by fabiobaltieri** [comment]
@Laczen just to be clear: your suggestion is to use multiple compatibles, basically adding one for the device type, and then iterate over those, right?

**Comment by Laczen** [comment]
> @Laczen just to be clear: your suggestion is to use multiple compatibles, basically adding one for the device type, and then iterate over those, right?

Yes, use a compatible to describe the device-type.

It is then possible to drive the creation of the linker section based upon `DT_HAS_XXX_ENABLED` and have the driver register to the appropiate linker section (the node has the  xxx-compatible). I don't think any iteration is needed.

**Comment by fabiobaltieri** [comment]
Yeah my point is that it would  be nice to use the api struct information instead, that way the system would just work.

**Comment by Laczen** [comment]
> Yeah my point is that it would be nice to use the api struct information instead rather than specifying it manually, that way the system would just work.

Using the api struct information could be done by modifying the `DEVICE_DT_INST_DEFINE()` macro, not registering a pointer but instead the api. This will take some macrobatics to support compile optimizations to reduce the api struct (some routines maybe missing depending on compile options). This would require quite some work.

Modifying DEVICE_DT_INST_DEFINE() to take a specific linker section (for storing the pointer to the devices api struct) should be quite simple and this would also allow checking if the dev belongs to a specific type (as suggested by @pdgendt). Maybe we can even just sort the devices based upon the specified compatible and retrieve the start and end of these devices types automatically. It would not be as optimal as directly registering the api structure but it are just a number of pointers to the api structs.

**Comment by henrikbrixandersen** [comment]
It's great you have all sorts of ideas how this could be done in different ways. I urge you to create PRs implementing this functionality as alternatives to this.

The suggestion about having a common device type be part of the compatibles list is not as easy as it sounds, though. All DTS(I) files would need to be modified for this solution, whereas the solution implemented here only requires three lines added to each base devicetree binding file.

**Comment by pdgendt** [comment]
> It's great you have all sorts of ideas how this could be done in different ways. I urge you to create PRs implementing this functionality as alternatives to this. This PR has been open for three months now without anybody stepping up with another proposed solution that does incur a run-time overhead in one way or another (ROM usage, CPU cycles, ...).

I think this PR is a nice addition, it solves the problem of listing devices at compile time, whereas the linker idea is useful for runtime checks.

**Comment by Laczen** [comment]
> It's great you have all sorts of ideas how this could be done in different ways. I urge you to create PRs implementing this functionality as alternatives to this. This PR has been open for three months now without anybody stepping up with another proposed solution that does incur a run-time overhead in one way or another (ROM usage, CPU cycles, ...).
> 
> The suggestion about having a common device type be part of the compatibles list is not as easy as it sounds, though. All DTS(I) files would need to be modified for this solution, whereas the solution implemented here only requires three lines added to each base devicetree binding file.

@henrikbrixandersen, all ideas that have been formulated where about extending the proposal. And these extensions could be done based upon a device-type or a compatible.

Regarding the use of compatibles versus device-type: The devicetree in linux used to have something like a device-type, but it has been deprecated so I find it strange that zephyr would need it. In zephyr we are already "grouping" items based upon compatibles, e.g. "fixed-partitions", "soc-nv-flash", so why take a different approach for devices ?

Regarding the need to modify all (DTS)I files or not, this is correct, the PR needs less work in (DTS)I files.

**Comment by de-nordic** [comment]
I like the idea, it very much reminds me of PCI classes for devices, which kinda represents subsystems that will use the device in the end.

Regarding the DTS usage: I think that the is the right place, even though it seems to be software configuration I do not think it actually is; notice that with PCI you could ask device for a class it serves, so basically you would ask it what subsystem it will be attached to - that is the same thing - but here DTS is what we have and we want to know at compile time.

Yes, you could probably do that with `_HAS_` Kconfigs, but these should be reserved for advertisement that device driver is available rather then the device, we already what devices we have from tree (if it is status "okay").

As far as I also understand: this is very automatic; once you derive your device binding from specific class it automatically advertises itself as a device of a given class. This is nice.

I think that this is nicely done and addresses the issue correctly; I would prefer "device-class" over "device-type", but I can live with the later.

Anyway, I like how this is done.

**Comment by henrikbrixandersen** [comment]
> I think that this is nicely done and addresses the issue correctly; I would prefer "device-class" over "device-type", but I can live with the later.

I started out with `zephyr,device-class` actually, and I wouldn't be opposed to reverting back to that.

File: dts/bindings/base/base.yaml [hunk 1]
```diff
@@ -3,6 +3,10 @@
 include: [pm.yaml]
 
 properties:
+  zephyr,device-type:
+    type: string
+    description: indicates the device type.
```
Comment (gmarull) [inline dts/bindings/base/base.yaml:8]: I'm sorry, but to me, using devicetree to add yet more software properties is a -1. Usage of `zephyr,` has become a mess to a point where we should likely rename `devicetree` to something else, maybe `ztree` or something. This solution also doesn't account for device nodes that can have >1 device type, which even if not supported now, they should if we ever pretend to fully align with Linux.

**Review by gmarull — CHANGES_REQUESTED** [review]

File: dts/bindings/base/base.yaml (hunk 1 above)
Comment (decsny) [reply dts/bindings/base/base.yaml:8]: adding that my suggestion just above would theoretically be able to handle multiple types

**Pull request closed without merging**